File: 1000~1999/Q1753.py
import sys
import heapq

INF = int(1e9)

V, E = map(int, sys.stdin.readline().split())
K = int(sys.stdin.readline())

graph = [[] for _ in range(V+1)] # 각 노드 연결된 정보 담는 그래프
distance = [INF] * (V+1) # 최단거리 담는 리스트

for i in range(E): # 간선 정보 입력받기
    u, v, w = map(int, sys.stdin.readline().split())
    graph[u].append((v, w))

def dijkstra(start):
    q = []
    heapq.heappush(q, (0, start))
    distance[start] = 0

    while q:
        dist, now = heapq.heappop(q)

        if distance[now] < dist:
            continue
    
        for i in graph[now]:
            cost = dist + i[1]
            
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(q, (cost, i[0]))

dijkstra(K)
for i in range(1, V+1):
    if distance[i] == INF:
        print("INF")
    else:
        print(distance[i])

'''
입력 예시
5 6
1
5 1 1
1 2 2
1 3 3
2 3 4
2 4 5
3 4 6

출력 예시
0
2
3
7
INF
'''

File: 1000~1999/Q1753.java
import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Q1753 {
    static int INF = Integer.MAX_VALUE;
    static ArrayList<Edge>[] graph;
    static int[] distance;
    static int v;

    static class Edge{
        int node;
        int dist;
        Edge(int node, int dist){
            this.node = node;
            this.dist = dist;
        }
    }

    static void dijkstra(int start){
        distance[start] = 0;
        PriorityQueue<Edge> queue = new PriorityQueue<>((o1, o2) -> o1.dist - o2.dist);
        queue.add(new Edge(start, 0));

        Edge edge;
        int now, dist;
        while(!queue.isEmpty()){
            edge = queue.poll();
            now = edge.node;
            dist = edge.dist;

            if(distance[now] < dist){
                continue;
            }

            for(Edge e: graph[now]){
                int cost = e.dist + dist;

                if(distance[e.node] > cost){
                    distance[e.node] = cost;
                    queue.add(new Edge(e.node, cost));
                }
            }
        }
        
        return;
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        v = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(br.readLine());

        graph = new ArrayList[v+1];
        distance = new int[v+1];
        Arrays.fill(distance, INF);

        for(int i=1; i<=v; i++){
            graph[i] = new ArrayList<>();
        }

        int a, b, c;
        for(int i=0; i<e; i++){
            st = new StringTokenizer(br.readLine());
            a = Integer.parseInt(st.nextToken());
            b = Integer.parseInt(st.nextToken());
            c = Integer.parseInt(st.nextToken());

            graph[a].add(new Edge(b, c));
        }

        dijkstra(k);

        StringBuilder sb = new StringBuilder();
        for(int i=1; i<=v; i++){
            if(distance[i] == INF){
                sb.append("INF\n");
            }
            else{
                sb.append(distance[i] + "\n");
            }
        }

        System.out.print(sb);
    }    
}